File: Programming/Lab_5/sort_module.cpp
#include "sort_module.h"
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <queue>
#include <cstdlib>

using namespace std;

static ProgressCallback _cb = nullptr;
static int _last = -1;
static void report(int p) {
    if (_cb && p != _last) { _cb(p); _last = p; }
}

struct Key {
    string s; int i = 0; double d = 0.0;
};

static bool less_by(const Key& a, const Key& b, const string& f) {
    if (f == "id" || f == "quantity") return a.i < b.i;
    if (f == "price") return a.d < b.d;
    return a.s < b.s;
}

// CSV
static Key csv_key(const string& line, const string& f) {
    Key k;
    size_t p1 = line.find(','), p2 = line.find(',', p1+1),
           p3 = line.find(',', p2+1), p4 = line.find(',', p3+1);
    if (f == "id") k.i = stoi(line.substr(0, p1));
    else if (f == "date") k.s = line.substr(p1+1, p2-p1-1);
    else if (f == "product") k.s = line.substr(p2+1, p3-p2-1);
    else if (f == "price") k.d = stod(line.substr(p3+1, p4-p3-1));
    else if (f == "quantity") k.i = stoi(line.substr(p4+1));
    return k;
}

DLL_EXPORT bool sort_csv_file(const char* field, int* rows, ProgressCallback cb) {
    _cb = cb; _last = -1; report(0);
	ifstream in("data.csv");
    if (!in) return false; // Проверка на существование файла (нет в ifstream, оставляем как есть)
    
    string header, line, f(field);
    getline(in, header);
    system("mkdir tmp 2>nul");
    
    vector<string> tmp_files;
    int total = 0;
    
    // ФАЗА 1
    while (true) {
        vector<pair<string, Key>> chunk;
        // Читаем чанки
        while (chunk.size() < 400000 && getline(in, line)) {
            if (!line.empty()) {
                chunk.push_back(make_pair(line, csv_key(line, f)));
            }
        }
        if (chunk.empty()) break;
        total += chunk.size();
        // Сортируем чанк по указанному полю
        sort(chunk.begin(), chunk.end(), [&](const pair<string,Key>& a, const pair<string,Key>& b) {
            return less_by(a.second, b.second, f);
        });
        string name = "tmp/chunk_" + to_string(tmp_files.size()) + ".tmp";
        // Записываем отсортированный чанк во временный текстовый файл
        ofstream out(name.c_str());
        for (size_t i = 0; i < chunk.size(); ++i) {
            out << chunk[i].first << "\n";
        }
        tmp_files.push_back(name);
        report((int)(tmp_files.size() * 50 / 77));
    }
    report(50);
    
    // ФАЗА 2
    ofstream out("sorted.txt"); // Открываем файл в режиме записи
    out << header << "\n";
    
    struct HeapItem {
        string line;
        Key key;
        int idx;
    };
    auto cmp = [&](const HeapItem& a, const HeapItem& b) {
        return !less_by(a.key, b.key, f);
    };
    priority_queue<HeapItem, vector<HeapItem>, decltype(cmp)> pq(cmp);
    
    vector<ifstream> files(tmp_files.size());
    // Открываем все temp файлы в режиме текстового чтения
    for (size_t i = 0; i < tmp_files.size(); ++i) {
        files[i].open(tmp_files[i].c_str());
        if (getline(files[i], line)) {
            HeapItem item;
            item.line = line;
            item.key = csv_key(line, f);
            item.idx = (int)i;
            pq.push(item);
        }
    }
    
    int merged = 0;
    while (!pq.empty()) {
        HeapItem top = pq.top(); pq.pop(); // Извлекаем наименьшую запись
        out << top.line << "\n"; // Записываем в результат
        if (++merged % 50000 == 0) report(50 + merged * 50 / total);
        line.clear();
        // Читаем следующую запись
        if (getline(files[top.idx], line)) {
            HeapItem item;
            item.line = line;
            item.key = csv_key(line, f);
            item.idx = top.idx;
            pq.push(item);
        }
        // Обновление прогресса каждые 50000 записей (вверху)
    }
    
    // Очистка после завершения сортировки
    for (size_t i = 0; i < files.size(); ++i) files[i].close();
    // Удаление временных файлов
    for (size_t i = 0; i < tmp_files.size(); ++i) remove(tmp_files[i].c_str());
    system("rmdir /s /q tmp 2>nul");
    report(100);
    *rows = total;
    return true;
}

// BIN
#pragma pack(push, 1)
struct BinRecord { int id, year, month, day; char product[50]; int price, quantity; };
#pragma pack(pop)

static Key bin_key(const BinRecord& r, const string& f) {
    Key k;
    if (f == "id") k.i = r.id;
    else if (f == "date") k.s = to_string(r.year) + "-" + to_string(r.month) + "-" + to_string(r.day);
    else if (f == "product") k.s = string(r.product);
    else if (f == "price") k.d = (double)r.price;
    else if (f == "quantity") k.i = r.quantity;
    return k;
}

DLL_EXPORT bool sort_bin_file(const char* field, int* rows, ProgressCallback cb) {
    _cb = cb; _last = -1; report(0);
    ifstream in("data.bin", ios::binary);
    if (!in) return false; // Проверка на существование файла
    
    system("mkdir tmp 2>nul");
    const size_t RS = sizeof(BinRecord);
    in.seekg(0, ios::end);
    size_t total = in.tellg() / RS; // Определяем размер
    in.seekg(0);
    
    vector<string> tmp_files;
    const size_t CHUNK = (10 * 1024 * 1024) / RS; // Ограничиваем память
    
    // ФАЗА 1
    while (true) {
        vector<pair<BinRecord, Key>> chunk;
        BinRecord r;
        // Читаем чанки
        for (size_t i = 0; i < CHUNK && in.read((char*)&r, RS); ++i) {
            chunk.push_back(make_pair(r, bin_key(r, field)));
        }
        if (chunk.empty()) break;
        // Сортируем чанк по указанному полю
        sort(chunk.begin(), chunk.end(), [&](const pair<BinRecord,Key>& a, const pair<BinRecord,Key>& b) {
            return less_by(a.second, b.second, field);
        });
        string name = "tmp/bin_" + to_string(tmp_files.size()) + ".tmp";
        // Записываем отсортированный чанк во временный бинарный файл
        ofstream out(name.c_str(), ios::binary);
        for (size_t i = 0; i < chunk.size(); ++i) {
            out.write((char*)&chunk[i].first, RS);
        }
        tmp_files.push_back(name);
        report((int)(tmp_files.size() * 50 / 103));
    }
    report(50);
    
    // ФАЗА 2
    struct BinHeapItem {
        BinRecord rec;
        Key key;
        int idx;
    };
    auto cmp = [&](const BinHeapItem& a, const BinHeapItem& b) {
        return !less_by(a.key, b.key, field);
    };
    priority_queue<BinHeapItem, vector<BinHeapItem>, decltype(cmp)> pq(cmp);
    
    vector<ifstream> files(tmp_files.size());
    // Открываем все temp файлы в режиме бинарного чтения
    for (size_t i = 0; i < tmp_files.size(); ++i) {
        files[i].open(tmp_files[i].c_str(), ios::binary);
        BinRecord r;
        if (files[i].read((char*)&r, RS)) {
            BinHeapItem item;
            item.rec = r;
            item.key = bin_key(r, field);
            item.idx = (int)i;
            pq.push(item);
        }
    }
    
    ofstream out("sorted.txt"); // Открываем файл в режиме записи
    out << "id,date,product,price,quantity\n";
    
    size_t merged = 0;
    while (!pq.empty()) {
        BinHeapItem top = pq.top(); pq.pop(); // Извлекаем наименьшую запись
        // Распаковываем запись
        out << top.rec.id << "," << top.rec.year << "-" << top.rec.month << "-" << top.rec.day << ","
            << top.rec.product << "," << top.rec.price << "," << top.rec.quantity << "\n"; // Записываем в CSV формате
        if (++merged % 100000 == 0) report(50 + (int)(merged * 50 / total));
        BinRecord r;
        // Читаем следующую запись
        if (files[top.idx].read((char*)&r, RS)) {
            BinHeapItem item;
            item.rec = r;
            item.key = bin_key(r, field);
            item.idx = top.idx;
            pq.push(item);
        }
        // Обновление прогресса каждые 100000 записей (вверху)
    }
    
    // Очистка после завершения сортировки
    for (size_t i = 0; i < files.size(); ++i) files[i].close();
    for (size_t i = 0; i < tmp_files.size(); ++i) remove(tmp_files[i].c_str());
    system("rmdir /s /q tmp 2>nul");
    report(100);
    *rows = (int)total;
    return true;
}

File: Programming/Lab_5/sort_module.py
import os
import heapq
import struct

# Прогресс
_cb = None
_last_progress = -1
def _report(p):
    global _last_progress
    if _cb and p != _last_progress:
        _cb(p)
        _last_progress = p

# CSV
def _make_csv_key(line, field):
    parts = line.split(",")
    if field == "id":
        return int(parts[0])
    elif field == "date":
        return parts[1]
    elif field == "product":
        return parts[2]
    elif field == "price":
        return float(parts[3])
    elif field == "quantity":
        return int(parts[4])
    return parts[0]

def sort_csv_file(field_name, progress_callback=None):
    global _cb, _last_progress
    _cb = progress_callback
    _last_progress = -1
    _report(0)
    if not os.path.exists("data.csv"): # Проверка на существование файла
        return False, 0
    os.makedirs("tmp", exist_ok=True)
    CHUNK = 400000 # Размер одной серии (строк)
    temp_files = []
    total = 0
    with open("data.csv", "r", encoding="utf-8") as f: # Открываем CSV в режиме текстового чтения
        header = f.readline()

        # ФАЗА 1
        while True:
            chunk = []
            # Читаем чанки
            for _ in range(CHUNK):
                line = f.readline()
                if not line:
                    break
                line = line.strip()
                if line:
                    chunk.append(line)
            if not chunk:
                break
            total += len(chunk)
            # Сортируем чанк по указанному полю
            chunk.sort(key=lambda x: _make_csv_key(x, field_name))
            name = f"tmp/chunk_{len(temp_files)}.tmp"
            # Записываем отсортированный чанк во временный текстовый файл
            with open(name, "w", encoding="utf-8") as out:
                out.write("\n".join(chunk) + "\n")
            temp_files.append(name)
            _report(int(len(temp_files) * 50 / 77))
    _report(50)

    # ФАЗА 2
    with open("sorted.txt", "w", encoding="utf-8") as out: # Открываем файл в режиме записи
        out.write(header)
        heap = []
        files = []
        # Открываем все temp файлы в режиме текстового чтения
        for i, fname in enumerate(temp_files):
            f = open(fname, "r", encoding="utf-8")
            files.append(f)
            line = f.readline().strip()
            if line:
                heapq.heappush(heap, (_make_csv_key(line, field_name), line, i))
        merged = 0
        while heap:
            key, line, idx = heapq.heappop(heap) # Извлекаем наименьшую запись
            out.write(line + "\n") # Записываем в результат
            merged += 1
            # Читаем следующую запись
            next_line = files[idx].readline().strip()
            if next_line:
                heapq.heappush(heap,
                    (_make_csv_key(next_line, field_name), next_line, idx))
            # Обновление прогресса каждые 50000 записей
            if merged % 50000 == 0:
                _report(50 + int(merged * 50 / total))
        # Очистка после завершения сортировки
        for f in files:
            f.close()
    # Удаление временных файлов
    for f in temp_files:
        os.remove(f)
    os.rmdir("tmp")
    _report(100)
    return True, total

# BIN
BIN_STRUCT = struct.Struct("<iiii50sii")

def _make_bin_key(record, field):
    id_, year, month, day, product, price, quantity = record
    product = product.decode("utf-8").rstrip("\x00")
    if field == "id":
        return id_
    elif field == "date":
        return f"{year}-{month}-{day}"
    elif field == "product":
        return product
    elif field == "price":
        return price
    elif field == "quantity":
        return quantity
    return id_

def sort_bin_file(field_name, progress_callback=None):
    global _cb, _last_progress
    _cb = progress_callback
    _last_progress = -1
    _report(0)
    if not os.path.exists("data.bin"): # Проверка на существование файла
        return False, 0
    os.makedirs("tmp", exist_ok=True)
    size = os.path.getsize("data.bin") # Определяем размер
    record_size = BIN_STRUCT.size
    total_records = size // record_size
    MAX_MEMORY = 10 * 1024 * 1024 # Ограничиваем память
    CHUNK_RECORDS = MAX_MEMORY // record_size
    temp_files = []

    # ФАЗА 1
    with open("data.bin", "rb") as f: # Открываем bin в режиме бинарного чтения
        while True:
            chunk = []
            # Читаем чанки
            for _ in range(CHUNK_RECORDS):
                data = f.read(record_size)
                if not data:
                    break
                chunk.append(BIN_STRUCT.unpack(data))
            if not chunk:
                break
            # Сортируем чанк по указанному полю
            chunk.sort(key=lambda r: _make_bin_key(r, field_name))
            name = f"tmp/bin_chunk_{len(temp_files)}.tmp"
            # Записываем отсортированый чанк во временный бинарный файл
            with open(name, "wb") as out:
                for record in chunk:
                    out.write(BIN_STRUCT.pack(*record))
            temp_files.append(name)
            _report(int(len(temp_files) * 50 / 103))
    _report(50)

    # ФАЗА 2
    heap = []
    files = []
    # Открываем все temp файлы в режиме бинарного чтения
    for i, fname in enumerate(temp_files):
        f = open(fname, "rb")
        files.append(f)
        data = f.read(record_size)
        if data:
            record = BIN_STRUCT.unpack(data)
            heapq.heappush(heap,
                (_make_bin_key(record, field_name), record, i))
            
    # Запись результата
    with open("sorted.txt", "w", encoding="utf-8") as out: # Открываем файл в режиме записи
        out.write("id,date,product,price,quantity\n")
        merged = 0
        while heap:
            key, record, idx = heapq.heappop(heap) # Извлекаем наименьшую запись
            id_, year, month, day, product, price, quantity = record
            product = product.decode("utf-8").rstrip("\x00") # Распаковываем запись
            out.write(f"{id_},{year}-{month}-{day},{product},{price},{quantity}\n") # Записываем в CSV формате
            merged += 1
            # Читаем следующую запись
            data = files[idx].read(record_size)
            if data:
                record = BIN_STRUCT.unpack(data)
                heapq.heappush(heap,
                    (_make_bin_key(record, field_name), record, idx))
            # Обновление прогресса каждые 100000 записей
            if merged % 100000 == 0:
                _report(50 + int(merged * 50 / total_records))

    # Очистка после завершения сортировки
    for f in files:
        f.close()
    for f in temp_files:
        os.remove(f)
    os.rmdir("tmp")
    _report(100)
    return True, total_records